### PR TITLE
[GPU][CPU] Support u2 (INT2) weight decompression: optimized FC kernel and fused MoE 3-GEMM

### PR DESCRIPTION
**Pull request by Cyberpunk1210**
### Details:

Adds u2 (2-bit unsigned, 4 values per byte) weight decompression to the GPU and CPU plugins.

**This PR is additive to the u2 support already in master** and does not re-add any of it:
 - #36170 added the `UINT2` data/weights types, `int2_utils.cl` unpack helpers,
   `convert_fc_to_compressed` handling and a u2 FC path on the **reference** kernel
   (`fully_connected_gpu_bfyx_ref`).
 - #37243 added the u2 **data** reorder (`reorder_data.cl` UINT2_INPUT/UINT2_OUTPUT,
   `ReorderKernelRef`'s UINT2 key, `convert_as_uint2_float`) and marked `ov::element::u2`
   supported in the GPU plugin.

On top of that, this PR adds:

**GPU**
 - u2 in the **optimized** `fully_connected_gpu_bf_tiled` FC kernel (both the default and the SLM
   path), so a u2 FC no longer falls back to the reference kernel.
 - u2 **weight** reorder in `reorder_weights_int4`: OIYX padding, IOYX->OIYX transpose and the
   `os_iyx_osv16` / `os_is_yx_osv64_isv2` repacks, gathering 4 values per output byte.
 - u2 compressed dictionaries in `gather_ref`.
 - A native u2 grouped GEMM in the fused MoE 3-GEMM prefill path, in two variants (DPAS/systolic on
   xe2+, FMA elsewhere), consuming the u2 expert weights directly. The alternative -- unpacking u2
   into u4 scratch so the existing u4 grouped GEMM can run -- needs 12.08 GB of scratch on
   Qwen3.6-35B-A3B; neither oneDNN's JIT GEMM generator (no 2-bit type, at most 2 elements per byte)
   nor the batched GEMV (one work-group per token/expert pair, so weights are re-read per token)
   is a workable substitute.
 - The u2 weight type plumbed through the two jitters and `fetch_weights.cl`.
 - OpenCL build logs are now carried in the thrown exception message rather than only in
   `GPU_DEBUG_INFO`, which is compiled out in release builds -- without this a kernel that fails to
   build reports only `clBuildProgram, error code: -11`.

**CPU**
 - u2 in compressed `Gather` (`execCompressed2Bit`), including the out-of-range row behaviour.

`GatherMatmul` is deliberately not extended to u2 on either plugin: u2 MoE layers are fused into
the `MOE` op and served by `moe_3gemm`, so such a kernel would not be reached by a model. On CPU it
could not be honoured anyway -- `dnnl_gathermatmul_executor` builds `weights_md` with
`format_tag::any`, oneDNN then selects `aBC4c24b4c` on MoE expert weight shapes, and no oneDNN
release has a u2 reorder for that tag. u2 rank-3 compressed MatMuls keep taking the CPU
FullyConnected path, which already supports u2. **This PR therefore has no oneDNN dependency and
needs no submodule bump.**

Motivation: run INT2-quantized MoE models (e.g. a 256-expert Qwen3.6-35B-A3B variant, routed-expert
weights at ~2.25 bit/param) without falling back to a slower per-GEMM path.

### Verification

 - Accuracy: the 256-expert INT2 variant (10.6 GB weights) reaches 72% on BFCL `multi_turn_base` in
   think mode, the deployment configuration -- on par with the INT4 baseline (~70%) at a much
   smaller footprint.
 - GPU-vs-CPU logits comparison on the real model, both decode and prefill, on real non-repeating
   text: cosine similarity 0.999+ against the CPU reference on both paths. Top-1 agreement dips to
   ~0.93 on prefill, which traces to near-tied logits at a few positions in natural text rather
   than a numerical error -- per-position cosine stays >= 0.997 throughout.
 - GPU unit suite on Arc B390 (sharded): 45974 tests, 0 failures in the u2, MoE, gather, reorder and
   compressed-FC areas.
 - Tests added across the transformation, CPU functional and GPU unit suites, covering each
   correctness fix in the series.

**Post-rebase status.** The series was rebased from f4ad686740 onto current master to pick up
#37243 and six weeks of other drift. It has since been rebuilt on Arc B390 / Windows (clean
build, no warnings introduced) and the u2, MoE, gather, reorder and compressed-FC unit tests
re-run there, with no failures on the u2 code paths. Full-suite validation is left to CI.

### Review notes

**On comment density.** The u2 kernels carry more explanatory comment than is usual for this
codebase (~18% of added lines). That is deliberate and concentrated on failure modes that are
silent rather than obvious, e.g. `intel_sub_group_f16_f16_matrix_mad_k16` compiles at the xe2+
default sub-group size of 32 and returns wrong results there; a subgroup block read of weights
overhangs an mmap'd weights buffer and faults with `CL_OUT_OF_RESOURCES`; a layer-wide "is the zp
scalar" flag would decode one expert GEMM with another's rule on a mixed INT2_SYM/INT2_ASYM layer.
Duplicated comment blocks and banner rules have been removed; the remaining comments are the
*why*. Happy to cut further wherever a reviewer finds them redundant.

### Tickets:
 - N/A

### AI Assistance:
 - AI assistance used: yes
 - Claude Opus 5 (Claude Code) was used to draft the kernel and transformation changes across the
   commits in this PR under close supervision, and to perform the rebase onto current master.
 - Human validation performed: incremental builds on Windows (MSBuild Release), the GPU unit suite,
   and GPU-vs-CPU logits correctness checks on the real target model as described above.

### CODE REVIEW

**Comment by p-durandin** [comment]
build_jenkins

File: src/core/src/op/moe.cpp [hunk 1]
```diff
@@ -54,8 +54,10 @@ void MOE::validate_and_infer_types() {
     for (size_t i = 4; i < base_inputs_count; i++) {
         const auto& ps = get_input_partial_shape(i);
         NODE_VALIDATION_CHECK(this, ps.is_static(), "Weights must have static shape.");
-        // Note: dynamic element type means empty zero point input
-        if (get_input_element_type(i) != ov::element::dynamic) {
+        // Note: dynamic element type means empty zero point input;
+        // a scalar (rank-0) input is a folded per-tensor zero point which
+        // has no num_experts dimension to check.
+        if (get_input_element_type(i) != ov::element::dynamic && ps.size() > 0) {
```
Comment (praasz) [inline src/core/src/op/moe.cpp:60]: @mmikolajcz @mitruska 
Could you check it?

File: src/core/reference/src/op/concat.cpp [hunk 2]
```diff
@@ -57,6 +57,8 @@ void concat(const std::vector<const char*>& args,
             size_t size = shape_sizes[in_index] / steps;
             if (elem_type == ov::element::u4 || elem_type == ov::element::i4)
                 size /= 2;
+            else if (elem_type == ov::element::u2)
```
Comment (praasz) [inline src/core/reference/src/op/concat.cpp:60]: Is not valid change and should be reverted, There is another PR which solve it.

**Review by praasz — CHANGES_REQUESTED** [review]
